p2p/src/authenticated/lookup/actors/tracker/actor.rs

```diff
@@ -0,0 +1,460 @@
+    context: E,
+
+    // ---------- Configuration ----------
+    /// The maximum number of peers in a set.
+    max_peer_set_size: usize,
+
+    // ---------- Message-Passing ----------
+    /// The mailbox for the actor.
+    receiver: mpsc::Receiver<Message<E, C::PublicKey>>,
+
+    // ---------- State ----------
+    /// Tracks peer sets and peer connectivity information.
+    directory: Directory<E, C::PublicKey>,
+
+    /// Maps a peer's public key to its mailbox.
+    /// Set when a peer connects and cleared when it is blocked or released.
+    peer_mailboxes: HashMap<C::PublicKey, peer::Mailbox>,
+}
+
+impl<E: Spawner + Rng + Clock + GClock + RuntimeMetrics, C: Signer> Actor<E, C> {
+    /// Create a new tracker [Actor] from the given `context` and `cfg`.
+    #[allow(clippy::type_complexity)]
+    pub fn new(
+        context: E,
+        cfg: Config<C>,
+    ) -> (Self, Mailbox<E, C::PublicKey>, Oracle<E, C::PublicKey>) {
+        // Sign my own information
+        let myself = (cfg.crypto.public_key(), cfg.address);
+
+        // General initialization
+        let directory_cfg = directory::Config {
+            max_sets: cfg.tracked_peer_sets,
+            rate_limit: cfg.allowed_connection_rate_per_peer,
+            allow_private_ips: cfg.allow_private_ips,
+        };
+
+        // Create the mailboxes
+        let (sender, receiver) = mpsc::channel(cfg.mailbox_size);
+        let mailbox = Mailbox::new(sender.clone());
+        let releaser = Releaser::new(sender.clone());
+        let oracle = Oracle::new(sender);
+
+        // Create the directory
+        let directory = Directory::init(context.clone(), myself, directory_cfg, releaser);
+
+        (
+            Self {
+                context,
+                max_peer_set_size: cfg.max_peer_set_size,
+                receiver,
+                directory,
+                peer_mailboxes: HashMap::new(),
+            },
+            mailbox,
+            oracle,
+        )
+    }
+
+    /// Start the actor and run it in the background.
+    pub fn start(mut self) -> Handle<()> {
+        self.context.spawn_ref()(self.run())
+    }
+
+    async fn run(mut self) {
+        while let Some(msg) = self.receiver.next().await {
+            match msg {
+                Message::Register { index, peers } => {
+                    // Ensure that peer set is not too large.
+                    // Panic since there is no way to recover from this.
+                    let len = peers.len();
+                    let max = self.max_peer_set_size;
+                    assert!(len <= max, "peer set too large: {} > {}", len, max);
+
+                    self.directory.add_set(index, peers);
+                }
+                Message::Connect {
+                    public_key,
+                    mut peer,
+                } => {
+                    // Kill if peer is not authorized
+                    if !self.directory.allowed(&public_key) {
```

Comment:
We do not kill peers if they are no longer in the directory (after registry).

This was the rationale for continuing to keep a "heartbeat" style message. Alternatively (given you introduced `peer_mailbox`), we could iterate through all peer mailboxes during `Register` and kill any peers that fail this check.

Reply:
I think the fix here should include a test that a register of a new set that drops a peer sends a kill.

Reply:
I'm not sure I understand your comment / the suggested change.